File: 241003/토끼와 경주/rabit-and-race.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_map>
#include <queue>
using namespace std;

struct Rabbit
{
    int pid;
    long long d;
    int x;
    int y;
    int jump;

    Rabbit(int pid_, long long d_) : pid(pid_), d(d_), x(1), y(1), jump(0) {}
};

struct HeapEntry
{
    int jump;
    int sum_xy;
    int x;
    int y;
    int pid;
};

struct CompareHeapEntry
{
    bool operator()(const HeapEntry &a, const HeapEntry &b) const
    {
        if (a.jump != b.jump)
            return a.jump > b.jump;
        if (a.sum_xy != b.sum_xy)
            return a.sum_xy > b.sum_xy;
        if (a.x != b.x)
            return a.x > b.x;
        if (a.y != b.y)
            return a.y > b.y;
        return a.pid > b.pid;
    }
};

pair<int, int> calc(int x, int y, long long d, int dx, int dy, int N, int M)
{
    int nx, ny;

    if (dx != 0)
    {
        int period = 2 * (N - 1);
        long long tmp = (x - 1) + (long long)dx * d;
        tmp %= period;
        if (tmp < 0)
            tmp += period;
        tmp = min((long long)tmp, (long long)(period - tmp));
        nx = tmp + 1;
    }
    else
    {
        nx = x;
    }

    if (dy != 0)
    {
        int period = 2 * (M - 1);
        long long temp = (y - 1) + (long long)dy * d;
        temp %= period;
        if (temp < 0)
            temp += period;
        temp = min((long long)temp, (long long)(period - temp));
        ny = temp + 1;
    }
    else
    {
        ny = y;
    }

    return {nx, ny};
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int Q;
    cin >> Q;

    unordered_map<int, Rabbit> rabbits;

    priority_queue<HeapEntry, vector<HeapEntry>, CompareHeapEntry> heap;

    unordered_map<int, long long> scores;

    int N = 0, M = 0;

    while (Q--)
    {
        int q;
        cin >> q;

        if (q == 100)
        {
            int P;
            cin >> N >> M >> P;
            for (int j = 0; j < P; ++j)
            {
                int pid;
                long long d;
                cin >> pid >> d;
                Rabbit rabbit(pid, d);
                rabbits.emplace(pid, rabbit);

                HeapEntry he;
                he.jump = rabbit.jump;
                he.sum_xy = rabbit.x + rabbit.y;
                he.x = rabbit.x;
                he.y = rabbit.y;
                he.pid = rabbit.pid;
                heap.push(he);
            }
        }
        else if (q == 200)
        {
            int K;
            long long S;
            cin >> K >> S;
            vector<tuple<long long, long long, long long, long long, int>> r_best;

            for (int j = 0; j < K; ++j)
            {
                if (heap.empty())
                    break;
                HeapEntry he = heap.top();
                heap.pop();
                int pid = he.pid;

                auto it = rabbits.find(pid);
                if (it == rabbits.end())
                    continue;
                Rabbit &rabbit = it->second;

                if (he.jump != rabbit.jump || he.sum_xy != (rabbit.x + rabbit.y) || he.x != rabbit.x || he.y != rabbit.y || he.pid != rabbit.pid)
                    continue;

                vector<pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
                vector<tuple<long long, long long, long long, long long, long long>> candi_vec;

                for (auto &[dx, dy] : directions)
                {
                    pair<int, int> new_pos = calc(rabbit.x, rabbit.y, rabbit.d, dx, dy, N, M);
                    int nx = new_pos.first;
                    int ny = new_pos.second;
                    candi_vec.emplace_back(-(long long)(nx + ny), -(long long)nx, -(long long)ny, (long long)nx, (long long)ny);
                }

                sort(candi_vec.begin(), candi_vec.end());

                auto chosen = candi_vec[0];
                long long nx = get<3>(chosen);
                long long ny = get<4>(chosen);

                rabbit.x = nx;
                rabbit.y = ny;
                rabbit.jump += 1;

                HeapEntry new_he;
                new_he.jump = rabbit.jump;
                new_he.sum_xy = rabbit.x + rabbit.y;
                new_he.x = rabbit.x;
                new_he.y = rabbit.y;
                new_he.pid = rabbit.pid;
                heap.push(new_he);

                long long score = nx + ny;
                for (auto &[o_pid, o_rabbit] : rabbits)
                {
                    if (o_pid != pid)
                    {
                        scores[o_pid] += score;
                    }
                }

                // Store this rabbit's data as a candidate for the best rabbit
                r_best.emplace_back(-(nx + ny), -nx, -ny, -(long long)pid, pid);
            }

            if (!r_best.empty())
            {
                sort(r_best.begin(), r_best.end());
                int best_pid = get<4>(r_best[0]);
                scores[best_pid] += S;
            }
        }
        else if (q == 300)
        {
            int pid_t;
            long long L;
            cin >> pid_t >> L;
            auto it = rabbits.find(pid_t);
            if (it != rabbits.end())
            {
                it->second.d *= L;
            }
        }
        else if (q == 400)
        {
            long long max_score = 0;
            for (auto &[pid, score] : scores)
            {
                if (score > max_score)
                    max_score = score;
            }
            cout << max_score << "\n";
        }
    }
}

File: 241003/토끼와 경주/rabit-and-race.py
import heapq
import sys
from collections import defaultdict

input = lambda: sys.stdin.readline().rstrip()



class Rabbit:
    def __init__(self, pid, d):
        self.pid = pid
        self.d = d
        self.x = 1  # 1-based
        self.y = 1
        self.jump = 0

    def get_priority(self):
        return self.jump, self.x + self.y, self.x, self.y, self.pid


def calc(x, y, d, dx, dy, N, M):
    if dx != 0:  # 행 방향 이동
        period = 2 * (N - 1)
        nx = x - 1 + dx * d
        nx %= period
        nx = min(nx, period - nx)
        nx += 1
    else:
        nx = x

    if dy != 0:  # 열 방향 이동
        period = 2 * (M - 1)
        ny = y - 1 + dy * d
        ny %= period
        ny = min(ny, period - ny)
        ny += 1
    else:
        ny = y

    return nx, ny


def main():
    Q = int(input())  # 명령어 수
    idx = 1

    rabbits = {}
    heap = []
    scores = defaultdict(int)
    selected_rabbits = set()
    N, M = 0, 0

    for _ in range(Q):
        q, *data = map(int, input().split())
        idx += 1

        if q == 100:
            N, M, P, *data = data
            for i in range(P):
                pid, d = data[i * 2], data[i * 2 + 1]
                rabbit = Rabbit(pid, d)
                rabbits[pid] = rabbit
                heapq.heappush(heap, (rabbit.get_priority(), pid))

        elif q == 200:
            K, S = data
            r_best = []

            for _ in range(K):
                if not heap:
                    break
                _, pid = heapq.heappop(heap)
                rabbit = rabbits[pid]
                selected_rabbits.add(pid)

                # 4방향으로 이동한 후의 후보 위치 계산
                candidates = []
                directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
                for dx, dy in directions:
                    final_x, final_y = calc(rabbit.x, rabbit.y, rabbit.d, dx, dy, N, M)
                    candidates.append((-(final_x + final_y), -final_x, -final_y, final_x, final_y))

                # 다음 이동 위치 결정
                candidates.sort()
                _, _, _, final_x, final_y = candidates[0]

                # 토끼의 위치와 점프 횟수 업데이트
                rabbit.x, rabbit.y = final_x, final_y
                rabbit.jump += 1

                # 힙에 다시 삽입 (변경된 우선순위 반영)
                heapq.heappush(heap, (rabbit.get_priority(), pid))

                # 현재 토끼를 제외한 모든 토끼의 점수 갱신
                score_increment = final_x + final_y
                for other_pid in rabbits:
                    if other_pid != pid:
                        scores[other_pid] += score_increment

                # 현재 이동한 토끼의 정보를 베스트 토끼 후보로 저장
                r_best.append((-(final_x + final_y), -final_x, -final_y, -pid, pid))

            # K번의 이동 후 베스트 토끼 점수 업데이트
            if r_best:
                r_best.sort()
                best_pid = r_best[0][4]
                scores[best_pid] += S

        elif q == 300:
            pid_t, L = data
            if pid_t in rabbits:
                rabbits[pid_t].d *= L

        elif q == 400:
            # Find the maximum score among all rabbits
            max_score = max(scores.values()) if scores else 0
            print(max_score)


if __name__ == "__main__":
    main()